docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,301 @@
+
+## Summary
+
+The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
+scaling on Cluster API deployed clusters for MachineSets and MachineDeployments. One feature
+that is missing from this integration is the ability to scale down to, and up from, a MachineSet
+or MachineDeployment with zero replicas.
+
+This proposal describes opt-in mechanisms whereby Cluster API users and infrastructure providers can define
+the specific resource requirements for each Infrastructure Machine Template they create. In situations
+where there are zero nodes in the node group, and thus the autoscaler does not have information
+about the nodes, the resource requests are utilized to predict the number of nodes needed. This
+information is only used by the autoscaler when it is scaling from zero.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated infrastructure providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and infrastructure providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Create an optional API contract in the Infrastructure Machine Template that allows infrastructure providers to specify
+  instance resource requirements that will be utilized by the cluster autoscaler.
+- Provide a mechanism for users to override the defined instance resource requirements on any given MachineSet or MachineDeployment.
+
+### Non-Goals/Future Work
+
+- Create an API contract that infrastructure providers must follow.
+- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
+- Support for MachinePools, either with the cluster autoscaler or using infrastructure provider native implementations (eg AWS AutoScalingGroups).
+- Create an autoscaling custom resource for Cluster API.
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for nodes within the target node group that will be scaled. The autoscaler
+uses this information to create a prediction about how many nodes should be created when scaling. In
+most situations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
```

Comment:
> When exactly does the autoscaler access this information? Only if there are exactly 0 replicas?

that is correct. 

> E.g. if a user changes the annotation, currently existing nodes should not be affected. Only if we are at 0 replicas the annotation will be read.

yes, in the situations where there are 1 or more replicas in the node group, it will use those nodes to predict the sizing of new nodes.